torch/csrc/jit/codegen/cuda/lower_mma.cpp

```diff
@@ -0,0 +1,609 @@
+    } else {
+      TORCH_INTERNAL_ASSERT(false, "unknown mma instruction");
+    }
+
+    return write_exprs;
+  }
+
+  std::vector<kir::Expr*> lowerMmaOperandSmemRead(kir::Expr* expr) {
+    std::vector<kir::Expr*> read_exprs;
+
+    auto in_tv = expr->inputs()[0]->as<kir::TensorIndex>()->view()->fuserTv();
+    TORCH_INTERNAL_ASSERT(in_tv->hasGemmParams());
+    auto& gemm_params = in_tv->gemmParams();
+
+    kir::IrBuilder ir_builder(GpuLower::current()->kernel());
+
+    bool is_read_a = in_tv == gemm_params.operand_a_smem_tv;
+    bool is_read_b = in_tv == gemm_params.operand_b_smem_tv;
+    TORCH_INTERNAL_ASSERT(
+        is_read_a || is_read_b, "MMA read tensor not a mma operand");
+
+    if (gemm_params.instruction_tile ==
+        mma_utils::GemmInstruction::M32N32K8_M8N8K4) {
+      auto orig_ti = expr->inputs()[0]->as<kir::TensorIndex>();
+      auto in_ti = ir_builder.create<kir::TensorIndex>(
+          orig_ti->view()->fuserTv(), orig_ti->indices(), orig_ti->strides());
+      in_ti->indexMode() = kir::TensorIndex::TensorIndexMode::Linear2D;
+      auto read_swizzle = kir::TensorIndex::TensorIndexSwizzle::NoSwizzle;
+
+      if (is_read_a) {
+        read_swizzle = kir::TensorIndex::TensorIndexSwizzle::MMA_M32N32K8ARead;
+      } else if (is_read_b) {
+        read_swizzle = kir::TensorIndex::TensorIndexSwizzle::MMA_M32N32K8BRead;
+      }
+
+      in_ti->indexSwizzle().push_back(read_swizzle);
+
+      // 2 vectorized loads need to be generated for a 32x32x8 instruction tile.
+      auto orig_set = expr->as<kir::UnaryOp>();
+      auto load_expr = ir_builder.create<kir::UnaryOp>(
+          UnaryOpType::Set, orig_set->out(), in_ti);
+      load_expr->forceVectorize(8);
+      read_exprs.push_back(load_expr);
+
+      // Add additional read exprs to cover the whole instruction tile
+      auto new_dest_reg_idx = ir_builder.create<kir::Int>(8);
+      auto new_dest_index = ir_builder.create<kir::TensorIndex>(
+          expr->outputs()[0]->as<kir::TensorIndex>()->view()->fuserTv(),
+          std::vector<kir::Val*>({new_dest_reg_idx}));
+
+      // The second load covers the second half of the row blocks
+      int row_offset = is_read_a ? gemm_params.instruction_tile_shape.m / 2
+                                 : gemm_params.instruction_tile_shape.k / 2;
+
+      auto new_row = ir_builder.addExpr(
+          ir_builder.create<kir::Int>(row_offset), in_ti->indices()[0]);
+
+      auto new_src_index = ir_builder.create<kir::TensorIndex>(
+          in_tv,
+          std::vector<kir::Val*>({new_row, in_ti->indices()[1]}),
+          in_ti->strides());
+      new_src_index->indexMode() = kir::TensorIndex::TensorIndexMode::Linear2D;
+      new_src_index->indexSwizzle().push_back(read_swizzle);
+
+      auto new_load_expr = ir_builder.create<kir::UnaryOp>(
+          UnaryOpType::Set, new_dest_index, new_src_index);
+      new_load_expr->forceVectorize(8);
+      read_exprs.push_back(new_load_expr);
```

Comment:
There are two things I was considering and I guess I prefer to keep them in fuser space at the moment:

One thinking is in our formal definition of swizzling and my thinking now is that swizzling is a function of index only and doesn't imply anything about the actual memory RW. So I was trying to avoid fused `swizzle+memory rw` in the resource strings even though we might still end up needing some for perf but again will be controllable by schedulers (not in this PR). After all `swizzle options` x `memory rw` can be quite a large space to build out in resource string. 

The other one is when we build out, these lowered ops will be different for different transposition and in Turing arch we might consider `ldmatrix`, which is also just a drop in replacement in this line.

I was hoping these are not too much complexity in terms of fusion lowering and I can add more comments and even package these logic better in utility functions.

Overall at the current starting point I just wanted to avoid combinatorically expand the resource string until it becomes too hard to do.